### PR TITLE
problem File

### PR DESCRIPTION
**Pull request by geniusgordon**
can upload files in edit problem

### CODE REVIEW

File: problem/forms.py [hunk 1]
```diff
@@ -23,21 +23,40 @@
...
+
+import autocomplete_light
+
+# create autocomplete interface and register
+class UserAutocomplete(autocomplete_light.AutocompleteModelBase):
+    search_fields = ['^username']
+    choices = User.objects.all()
+    model = User
+    attrs = {
+        'placeholder': '',
```
Comment (henryyang42) [inline problem/forms.py]: If you don't want to use placeholder, just delete it.

File: problem/static/problem/js/edit.js [hunk 2]
```diff
@@ -136,16 +137,21 @@ function hide_field() {
   $("#id_partial_judge_code").parent().hide();
   $("#id_special_judge_code").parent().hide();
 }
-$("select").on("change", function(e) {
-    hide_field();
-    if (this.value == "ERR_TORRENT")
+
+function show_field(option) {
+    if (option == "ERR_TORRENT")
```
Comment (nickelcheng) [inline problem/static/problem/js/edit.js]: Please fix typo: ```TORRENT``` -> ```TOLERANT```

File: problem/templates/problem/editTab/info.html [hunk 3]
```diff
@@ -38,6 +38,12 @@
     </div>
   {% endif %}
 {% endfor %}
+{% if has_special_judge_code %}
+Special Judge Code: <a href="/{{path.SPECIAL_PATH}}{{pid}}.c">{{pid}}.c</a>
```
Comment (nickelcheng) [inline problem/templates/problem/editTab/info.html]: Please consider .cpp sources

File: problem/templates/problem/editTab/info.html [hunk 4]
```diff
@@ -38,6 +38,12 @@
...
+{% endif %}
+{% if has_partial_judge_code %}
+Partial Judge Code: <a href="/{{path.PARTIAL_PATH}}{{pid}}.c">{{pid}}.c</a>
```
Comment (nickelcheng) [inline problem/templates/problem/editTab/info.html]: Ditto

File: problem/views.py [hunk 5]
```diff
@@ -102,6 +106,14 @@ def edit(request, pid=None):
             problem.sample_in = request.POST['sample_in']
             problem.sample_out = request.POST['sample_out']
             problem.save()
+            if "special_judge_code" in request.FILES:
+                with open('%s%s.c' % (SPECIAL_PATH, problem.pk), 'w') as t_in:
```
Comment (nickelcheng) [inline problem/views.py]: Ditto.

File: problem/views.py [hunk 6]
```diff
@@ -102,6 +106,14 @@ def edit(request, pid=None):
...
+                    for chunk in request.FILES['special_judge_code'].chunks():
+                        t_in.write(chunk)
+            if "partial_judge_code" in request.FILES:
+                with open('%s%s.c' % (PARTIAL_PATH, problem.pk), 'w') as t_in:
```
Comment (nickelcheng) [inline problem/views.py]: Ditto.

File: problem/views.py [hunk 7]
```diff
@@ -178,6 +196,14 @@ def testcase(request, pid, tid=None):
             testcase.memory_limit = request.POST['memory_limit']
             testcase.save()
             logger.info("testcase saved, tid = %s" % (testcase.pk))
+        if 't_in' in request.FILES:
+            TESTCASE_PATH = config_info.get_config('path', 'testcase_path')
+            with open('%s%s.in' % (TESTCASE_PATH, testcase.pk), 'w') as t_in:
+                for chunk in request.FILES['t_in'].chunks():
+                    t_in.write(chunk)
+            with open('%s%s.out' % (TESTCASE_PATH, testcase.pk), 'w') as t_out:
```
Comment (nickelcheng) [inline problem/views.py]: Does it need to check whether ```'t_out'``` exist or not?

Comment (geniusgordon) [inline problem/views.py]: can't submit if t_out is empty

**Comment by geniusgordon** [comment]
done

File: problem/views.py (hunk 7 above)
Comment (nickelcheng) [inline problem/views.py]: ok

**Comment by nickelcheng** [comment]
I got IOError
screenshot from 2015-04-04 23 09 58
Maybe you should modify ```install.py``` as well.

**Comment by bruce3557** [comment]
+1 if error is fixed

**Comment by nickelcheng** [comment]
Error had been fixed.
However, I can not download / view the uploaded testcase.
I can not find uploaded file in my file system, either.
BTW, ```nthuoj.cfg``` doesn't in your repo, I wonder how can it work without that file. Please merge the latest code.

**Comment by geniusgordon** [comment]
merged latest code

**Comment by nickelcheng** [comment]
1. Please auto refresh after add a new testcase.
2. When I click ```xxx.in``` or ```xxx.out```, I can not view what I uploaded. Also, the file didn't appear in my file system.

File: problem/views.py [hunk 8]
```diff
@@ -70,6 +71,9 @@ def detail(request, pid):
 
 @login_required
 def edit(request, pid=None):
+    SPECIAL_PATH = config_info.get_config('path', 'special_judge_path')
```
Comment (henryyang42) [inline problem/views.py]: Can you move that out of the function?
Otherwise, it will query config file each time you edit.

Comment (henryyang42) [inline problem/views.py]: These lines have appeared in `` problem/problem_info.py``.
Just reuse them.

File: problem/templates/problem/editTab/testcase.html [hunk 9]
```diff
@@ -11,10 +11,10 @@
   {% for t in testcase %}
   <tr data-target="{{t.pk}}">
     <td>
-      <a href="">{{t.pk}}.in</a>
+      <a href="/{{path.TESTCASE_PATH}}{{t.pk}}.in">{{t.pk}}.in</a>
```
Comment (henryyang42) [inline problem/templates/problem/editTab/testcase.html]: Are you sure django serve downloadable file that way?

**Comment by nickelcheng** [comment]
1. Please create folders for testcase and judge code in ```install.py```.
2. You should handle file upload exception.
3. ReUpload testcase doesn't work (I got ```"POST /problem/pid/testcase/0/ HTTP/1.1" 404 6447``` on console)
4. No button for uploading judge code QQ
screenshot from 2015-04-06 22 26 47

File: problem/forms.py [hunk 10]
```diff
@@ -47,7 +47,9 @@ class Meta:
             'pname',
             'owner',
             'visible',
+            'judge_language',
             'judge_source',
+            'judge_type',
             'error_torrence',
```
Comment (nickelcheng) [inline problem/forms.py]: I think this column should be changed to ```error_tolerance```.
Please modify it in both ```forms.py``` and ```models.py```.

**Comment by nickelcheng** [comment]
+1 besides comments.

File: problem/forms.py (hunk 10 above)
Comment (geniusgordon) [inline problem/forms.py]: judge code is uploaded when save problem

**Comment by geniusgordon** [comment]
add testcase log

**Comment by bruce3557** [comment]
Fine with this PR, please merge latest code

**Pull request merged**